lib/beautiful/colorful.lua

```diff
@@ -0,0 +1,1395 @@
+    return nil
+end
+
+-- }}} String
+
+-- {{{ Decimal
+
+--- Create a new color instance from an RGB decimal value.
+-- Will bitshift decimal 2 bytes left and call `color.from.hex4`
+-- @tparam number hex A decimal describing an RGB color
+-- @return A new color object, or nil if invalid
+-- @usage local c = colorful.color.from.hex3(0xffcc99)
+function color.from.hex3(hex)
+    return color.from.hex4(lshift(hex, 8) + 0xff)
+end
+
+--- Create a new color instance from an RGBA decimal value.
+-- @tparam number hex A decimal describing an RGBA color
+-- @return A new color object, or nil if invalid
+-- @usage local c = colorful.color.from.hex4(0xffcc99ff)
+function color.from.hex4(hex)
+    -- shift right 24 bits, 0s shifted in
+    local r = rshift(hex, 24)
+    -- clear upper bits as needed, and shift remaining bits down
+    local g = rshift(lshift(hex, 8), 24)
+    local b = rshift(lshift(hex, 16), 24)
+    local a = rshift(lshift(hex, 24), 24)
+    return color.from.rgba(r,g,b,a)
+end
+
+-- }}} Decimal
+
+-- {{{ RGB
+
+--- Create a new color instance from RGBA values.
+-- @tparam number r Red decimal value 0-255
+-- @tparam number g Green decimal value 0-255
+-- @tparam number b Blue decimal value 0-255
+-- @tparam number a Alpha decimal value 0-255
+-- @return A new color object, or nil if invalid
+-- @usage local c = colorful.color.from.rgba(255, 204, 153, 255)
+function color.from.rgba(r,g,b,a)
+    -- put r,g,b,a into [0,1]
+    r = clamp(r, 0, 255) / 255
+    g = clamp(g, 0, 255) / 255
+    b = clamp(b, 0, 255) / 255
+    a = clamp(a, 0, 255) / 255
+    -- make sure values are valid
+    if r == nil or g == nil or b == nil or a == nil then return nil end
+    -- hsl function expects RGB in range [0,1]
+    local h,s,l = rgb_to_hsl(r,g,b)
+    return create_color_object(r,g,b,h,s,l,a)
+end
+
+--- Create a new color instance from RGB values.
+-- Will call `color.from.rgba` with an Alpha value of 255
+-- @tparam number r Red decimal value 0-255
+-- @tparam number g Green decimal value 0-255
+-- @tparam number b Blue decimal value 0-255
+-- @return A new color object, or nil if invalid
+-- @usage local c = colorful.color.from.rgb(255, 204, 153)
+function color.from.rgb(r,g,b)
+    return color.from.rgba(r,g,b,255)
+end
+
+-- }}} RGB
+
+-- {{{ HSL
+
+--- Create a new color instance from HSLA values.
+-- @tparam number h Hue decimal value 0-360
+-- @tparam number s Saturation float value 0-1
+-- @tparam number l Lightness float value 0-1
+-- @tparam number a Alpha float value 0-1
+-- @return A new color object, or nil if invalid
+-- @usage local c = colorful.color.from.hsla(30, 0.8, 0.6, 1)
+function color.from.hsla(h,s,l,a)
+    -- hue is adjustable as it's degrees
+    if not between(h, 0, 360) then
+        while h < 0 do
+            h = h + 360
+        end
+        while h > 360 do
+            h = h - 360
+        end
+    end
+    s = clamp(s, 0.0, 1.0)
+    l = clamp(l, 0.0, 1.0)
+    if s == nil or l == nil then return nil end
```

Comment:
I think this is dead code. If one of them is `nil`, then `clamp` or `between` would have errored out. (Same for `from.rgba`)

Reply:
You're right. I left it there as a holdover before I changed it to use `clamp`. Can be removed.